src/main/java/org/springframework/data/couchbase/repository/query/CouchbaseQueryExecution.java

```diff
@@ -0,0 +1,305 @@
+	 *
+	 * @author Oliver Gierke
+	 * @author Mark Paluch
+	 * @author Christoph Strobl
+	 */
+	final class PagedExecution<FindWithQuery> implements CouchbaseQueryExecution {
+
+		private final ExecutableFindByQueryOperation.ExecutableFindByQuery<?> operation;
+		private final Pageable pageable;
+
+		public PagedExecution(ExecutableFindByQueryOperation.ExecutableFindByQuery<?> operation, Pageable pageable) {
+
+			Assert.notNull(operation, "Operation must not be null!");
+			Assert.notNull(pageable, "Pageable must not be null!");
+
+			this.operation = operation;
+			this.pageable = pageable;
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.couchbase.repository.query.CouchbaseQueryExecution#execute(org.springframework.data.couchbase.core.query.Query)
+		 */
+		@Override
+		public Object execute(Query query, Class<?> type, String collection) {
+
+			int overallLimit = query.getLimit();
+
+			ExecutableFindByQueryOperation.TerminatingFindByQuery<?> matching = operation.matching(query);
+
+			// Apply raw pagination
+			query.with(pageable);
+
+			// Adjust limit if page would exceed the overall limit
+			if (overallLimit != 0 && pageable.getOffset() + pageable.getPageSize() > overallLimit) {
+				query.limit((int) (overallLimit - pageable.getOffset()));
+			}
+
+			return PageableExecutionUtils.getPage(matching.all(), pageable, () -> {
+
+				long count = operation.matching(query.skip(-1).limit(-1)).count();
+				return overallLimit != 0 ? Math.min(count, overallLimit) : count;
+			});
+		}
+	}
+
+	/*
+		final class ReactivePagedExecution<FindWithQuery> implements CouchbaseQueryExecution { // TODO ??
+	
+			private final ReactiveFindByQueryOperation.ReactiveFindByQuery<?> operation;
+			private final Pageable pageable;
+	
+			public ReactivePagedExecution(ReactiveFindByQueryOperation.ReactiveFindByQuery<?> operation, Pageable pageable) {
+	
+				Assert.notNull(operation, "Operation must not be null!");
+				Assert.notNull(pageable, "Pageable must not be null!");
+	
+				this.operation = operation;
+				this.pageable = pageable;
+			}
+	
+	
+			/xx*
+			 * (non-Javadoc)
+			 * @see org.springframework.data.couchbase.repository.query.CouchbaseQueryExecution#execute(org.springframework.data.couchbase.core.query.Query)
+			 *xx/
+			@Override
+			public Object execute(Query query) {
+	
+				int overallLimit = query.getLimit();
+	
+				ReactiveFindByQueryOperation.TerminatingFindByQuery<?> matching = operation.matching(query);
+	
+				// Apply raw pagination
+				query.with(pageable);
+	
+				// Adjust limit if page would exceed the overall limit
+				if (overallLimit != 0 && pageable.getOffset() + pageable.getPageSize() > overallLimit) {
+					query.limit((int) (overallLimit - pageable.getOffset()));
+				}
+	
+				return PageableExecutionUtils.getPage(matching.all().collectList().block(), pageable, () -> {
+	
+					long count = operation.matching(query.skip(-1).limit(-1)).count().block();
```

Comment:
While the code isn't active, please note that calling `.block()` methods is almost a guarantee to break the entire processing flow. Please refrain from calling `.block()`.